src/dataflow-types/src/client.rs

```diff
@@ -1143,6 +1369,177 @@ pub mod tcp {
+                }
+            }
+        }
+
+        pub async fn recv(&mut self) -> Result<ComputeCommand, TryFromProtoError> {
+            loop {
+                match self.shared.queue.lock().await.as_mut() {
+                    Some(x) => match x.0.recv().await {
+                        Some(x) => return x.into_rust(),
+                        None => {
+                            println!("Recv returned None, endpoint gone, retry");
+                        }
+                    },
+
+                    // Other end absent, wait for connection
+                    None => {
+                        info!("recv called while no coordinator connected, waiting for coordinator connection.");
+                    }
+                }
+
+                self.shared.queue_change.notified().await;
+            }
+        }
+    }
+
+    type ResponseStream =
+        Pin<Box<dyn Stream<Item = Result<super::ProtoComputeResponse, tonic::Status>> + Send>>;
+
+    #[tonic::async_trait]
+    impl super::proto_compute_server::ProtoCompute for GrpcComputedServer {
+        type HandleStream = ResponseStream;
+
+        async fn handle(
+            &self,
+            req: Request<Streaming<super::ProtoComputeCommand>>,
+        ) -> Result<tonic::Response<Self::HandleStream>, tonic::Status> {
+            info!("GrpcComputedServer: remote client connected");
+
+            let mut in_stream = req.into_inner();
+            // The channel sizes are arbitrary, but it should be a
+            // small bounded channel such that backpressure is applied early.
+            let (cmd_tx, cmd_rx) = mpsc::channel(4);
+            let (resp_tx, resp_rx) = mpsc::channel(4);
+
+            // this spawn here is required if you want to handle connection error.
+            // If we just map `in_stream` and write it back as `out_stream` the `out_stream`
+            // will be drooped when connection error occurs and error will never be propagated
+            // to mapped version of `in_stream`.
+            let linger = self.linger;
+            let shutdown_tx = self.shutdown_tx.clone();
+            mz_ore::task::spawn(|| "ProtoComputeServer: ErrorChecker", async move {
+                while let Some(result) = in_stream.next().await {
+                    match result {
+                        Ok(v) => cmd_tx.send(v).await.expect("working rx"),
+
+                        Err(e) => {
+                            info!("Client disconnected with error {}", e);
+                            break;
+                        }
+                    }
+                }
+
+                if !linger {
+                    shutdown_tx.send(()).await.expect("working shutdown rx");
+                }
+                info!("Client disconnected");
+            });
+
+            // Store channels in state
+            *self.shared.queue.lock().await = Some((cmd_rx, resp_tx));
+            self.shared.queue_change.notify_waiters();
```

Comment:
what does this do? this is just a question based on my ignorance of how the boundary works

Reply:
I added a comment on top of the struct:

```
    /// If an error occurs, the consumer receives an error from the recv call. If no client is
    /// connected recv will block until a client is available. To implement the "waiting for
    /// client" the queue_change notification is used. recv will check first if a client is connected using
    /// queue. If queue is None, no client is connected and recv will await on the queue_change notification.
    /// The server does this vice-versa. Upon connection of a client, it will insert the endpoints
    /// into queue and trigger the queue_change, which will wake up the waiting clients.
    ```